File: duckdata/crosslink/python/shared_memory/shared_memory.py
"""
Shared memory operations for CrossLink
"""
import warnings

# Try to import C++ bindings
try:
    from .cpp_wrapper import is_cpp_available, CrossLinkCpp
    _cpp_available = is_cpp_available
except ImportError:
    _cpp_available = False
    
# Try to import DuckDB for fallback
try:
    import duckdb
    _duckdb_available = True
except ImportError:
    _duckdb_available = False

# Global cache for shared memory regions
_shared_memory_regions = {}

# Check if shared memory is available
def _check_shared_memory_available():
    # First check for C++ bindings which provide shared memory
    if _cpp_available:
        return True
    
    # Fall back to Python shared memory
    try:
        from multiprocessing import shared_memory
        return True
    except ImportError:
        return False

# Global variable to track if shared memory is available
_shared_memory_available = _check_shared_memory_available()

def setup_shared_memory(cl, arrow_table, dataset_id):
    """Set up shared memory for an Arrow table
    
    Args:
        cl: CrossLink instance
        arrow_table: PyArrow Table to share
        dataset_id: Dataset ID for naming
        
    Returns:
        Dict with shared memory information or None if not supported
    """
    # If C++ bindings are available, use them for optimal performance
    if _cpp_available and hasattr(cl, '_cpp_instance'):
        try:
            # Already using C++ implementation
            return cl._cpp_instance.setup_shared_memory(arrow_table, dataset_id)
        except Exception as e:
            cl._log(f"Failed to setup shared memory via C++: {e}", level="warning")
            # Continue to Python implementation as fallback
    
    # Python implementation (existing code)
    if not _shared_memory_available:
        cl._log("Shared memory requires Python 3.8+ with multiprocessing.shared_memory", level="warning")
        return None
        
    try:
        # Lazy import
        import pyarrow as pa
        from multiprocessing import shared_memory
        
        # Serialize the Arrow table to bytes
        sink = pa.BufferOutputStream()
        with pa.ipc.new_stream(sink, arrow_table.schema) as writer:
            writer.write_table(arrow_table)
        buffer = sink.getvalue()
        buffer_size = len(buffer.to_pybytes())
        
        # Create shared memory region
        shm_name = f"crosslink_{dataset_id}"
        shm = shared_memory.SharedMemory(name=shm_name, create=True, size=buffer_size)
        
        # Copy Arrow data to shared memory
        shm.buf[:buffer_size] = buffer.to_pybytes()
        
        # Track for cleanup
        cl._shared_memory_regions.append(shm)
        
        # Also store in global cache for cross-instance access
        global _shared_memory_regions
        _shared_memory_regions[shm_name] = shm
        
        return {
            "shared_memory_key": shm_name,
            "shared_memory_size": buffer_size,
            "arrow_schema": arrow_table.schema.serialize().hex()
        }
    except ImportError as e:
        cl._log(f"SharedMemory import failed: {e}", level="warning")
        return None
    except Exception as e:
        cl._log(f"Failed to setup shared memory: {e}", level="error")
        return None

def get_from_shared_memory(cl, shared_memory_key, metadata=None):
    """Get Arrow table from shared memory
    
    Args:
        cl: CrossLink instance
        shared_memory_key: Shared memory region name
        metadata: Dataset metadata with schema information
        
    Returns:
        Arrow table or None if failed
    """
    # Handle None or empty shared memory key
    if shared_memory_key is None or shared_memory_key == "":
        cl._log("Shared memory key is missing or empty, will fall back to standard queries", level="info")
        return None
        
    # If C++ bindings are available, use them for optimal performance
    if _cpp_available and hasattr(cl, '_cpp_instance'):
        try:
            # Already using C++ implementation
            result = cl._cpp_instance.get_from_shared_memory(shared_memory_key, metadata)
            if result is not None:
                return result
            cl._log(f"C++ shared memory lookup failed for key: {shared_memory_key}, will fall back to standard queries", level="info")
        except Exception as e:
            cl._log(f"Failed to access shared memory via C++: {e}, will fall back to standard queries", level="warning")
    
    # Python implementation (existing code)
    if not _shared_memory_available:
        cl._log("Shared memory not available, will fall back to standard queries", level="info")
        return None
        
    try:
        # Lazy import
        import pyarrow as pa
        from multiprocessing import shared_memory
        
        # Try to attach to shared memory region
        if shared_memory_key in _shared_memory_regions:
            # Use cached region
            shm = _shared_memory_regions[shared_memory_key]
        else:
            try:
                # Attach to existing region
                shm = shared_memory.SharedMemory(name=shared_memory_key, create=False)
                _shared_memory_regions[shared_memory_key] = shm
            except FileNotFoundError:
                cl._log(f"Shared memory region not found: {shared_memory_key}, will fall back to standard queries", level="info")
                return None
        
        # Create buffer from shared memory
        buffer = pa.py_buffer(shm.buf)
        
        # Try to extract schema from metadata
        if metadata and metadata.get('arrow_schema'):
            schema_hex = metadata['arrow_schema']
            if isinstance(schema_hex, dict) and 'serialized' in schema_hex:
                schema_hex = schema_hex['serialized']
                
            if isinstance(schema_hex, str):
                # Deserialize schema
                schema_bytes = bytes.fromhex(schema_hex)
                schema = pa.ipc.read_schema(pa.py_buffer(schema_bytes))
                
                # Read with schema
                reader = pa.ipc.RecordBatchStreamReader(buffer, schema)
                return reader.read_all()
        
        # Fallback to reading without schema
        reader = pa.ipc.open_stream(buffer)
        return reader.read_all()
        
    except Exception as e:
        cl._log(f"Failed to access shared memory: {e}, will fall back to standard queries", level="warning")
        return None

def fallback_to_duckdb(cl, identifier):
    """Fallback to using DuckDB when shared memory access fails
    
    Args:
        cl: CrossLink instance
        identifier: Dataset identifier
        
    Returns:
        Arrow table or None if failed
    """
    if not _duckdb_available:
        cl._log("DuckDB not available for fallback query", level="warning")
        return None
        
    try:
        import pyarrow as pa
        
        # Connect to DuckDB
        conn = duckdb.connect(cl.db_path)
        
        # Safety: clean identifier to prevent SQL injection
        safe_identifier = ''.join(c for c in identifier if c.isalnum() or c == '_')
        
        # Query the data
        query = f"SELECT * FROM {safe_identifier}"
        arrow_result = conn.execute(query).fetch_arrow_table()
        
        # Close connection
        conn.close()
        
        return arrow_result
    except Exception as e:
        cl._log(f"Failed to execute DuckDB fallback query: {e}", level="error")
        return None

def safe_shared_memory_get(cl, shared_memory_key, identifier, metadata=None):
    """Safely get data from shared memory with fallback to DuckDB
    
    Args:
        cl: CrossLink instance
        shared_memory_key: Shared memory key
        identifier: Dataset identifier for fallback
        metadata: Dataset metadata
        
    Returns:
        Arrow table or None if all methods fail
    """
    # Try shared memory first
    result = get_from_shared_memory(cl, shared_memory_key, metadata)
    
    # If shared memory access fails, fall back to DuckDB
    if result is None:
        cl._log(f"Shared memory access failed for {identifier}, falling back to DuckDB", level="info")
        result = fallback_to_duckdb(cl, identifier)
    
    return result

def cleanup_all_shared_memory():
    """Clean up all shared memory regions used by CrossLink
    
    This is a static method that can be called to clean up shared memory resources
    even if you don't have a reference to the CrossLink instance.
    
    Returns:
        Number of shared memory regions cleaned up
    """
    # If C++ bindings are available, use them for optimal performance
    if _cpp_available:
        try:
            # Create a temporary C++ instance to call cleanup
            cpp_instance = CrossLinkCpp()
            count = cpp_instance.cleanup_shared_memory()
            return count
        except Exception as e:
            warnings.warn(f"Failed to clean up shared memory via C++: {e}")
            # Fall back to Python implementation
    
    # Python implementation (existing code)
    if not _shared_memory_available:
        return 0
        
    from multiprocessing import shared_memory
    import re
    
    count = 0
    pattern = re.compile(r'^crosslink_[0-9a-f\-]+$')
    
    # Get all shared memory regions from the system
    try:
        # Find all shared memory segments that match our pattern
        for shm_name in [name for name in shared_memory.ShareableList.get_shared_memory_tracker().get_names() 
                       if pattern.match(name)]:
            try:
                # Open and unlink shared memory
                shm = shared_memory.SharedMemory(name=shm_name, create=False)
                shm.close()
                shm.unlink()
                count += 1
            except Exception:
                pass
    except Exception:
        # Fallback to cleaning only known regions
        global _shared_memory_regions
        for name, shm in list(_shared_memory_regions.items()):
            try:
                shm.close()
                shm.unlink()
                del _shared_memory_regions[name]
                count += 1
            except Exception:
                pass
    
    # Clear the global cache
    _shared_memory_regions.clear()
    
    return count

# Factory function to create appropriate implementation based on availability
def create_shared_memory_manager(cl):
    """Create a shared memory manager based on available implementations
    
    Args:
        cl: CrossLink instance
        
    Returns:
        A shared memory manager instance
    """
    if _cpp_available:
        try:
            from .cpp_wrapper import CrossLinkCpp
            return CrossLinkCpp(cl.db_path, cl._debug)
        except Exception as e:
            warnings.warn(f"Failed to create C++ shared memory manager: {e}")
    
    # Fall back to Python implementation
    return None 

File: duckdata/crosslink/julia/shared_memory/shared_memory.jl
# Import necessary modules at top level
using SharedArrays

# Flag to indicate whether shared memory is available
global _SHARED_MEMORY_AVAILABLE = false

# Preferred shared memory module to use
global _SHARED_MEMORY_MODULE = :None

# Dictionary of known shared memory regions
global _SHARED_MEMORY_REGIONS = Dict{String, String}()

# Initialize shared memory capabilities
function init_shared_memory()
    global _SHARED_MEMORY_AVAILABLE = false
    global _SHARED_MEMORY_MODULE = :None
    
    # Try to detect available shared memory modules
    try
        # If C++ wrapper is available, use it
        if @isdefined(CrossLinkCppWrapper) && CrossLinkCppWrapper.has_shared_memory()
            _SHARED_MEMORY_MODULE = :CppWrapper
            _SHARED_MEMORY_AVAILABLE = true
            return true
        end
    catch e
        # Continue to other options
        @warn "C++ wrapper not available for shared memory: $e"
    end
    
    # Try to use SharedArrays
    try
        _SHARED_MEMORY_MODULE = :SharedArrays
        _SHARED_MEMORY_AVAILABLE = true
        return true
    catch e
        # Continue to other options
        @warn "SharedArrays not available for shared memory: $e"
    end
    
    # If all else fails, use memory mapping
    try
        _SHARED_MEMORY_MODULE = :Mmap
        _SHARED_MEMORY_AVAILABLE = true
        return true
    catch e
        # Unable to initialize shared memory
        _SHARED_MEMORY_AVAILABLE = false
        @warn "Memory mapping not available for shared memory: $e"
        return false
    end
end

# Initialize on module load
init_shared_memory()

"""
    check_shared_memory_available()

Check if shared memory is available in the Julia environment.
"""
function check_shared_memory_available()
    try
        # In a production environment, you'd check for proper shared memory support
        # For this implementation, we'll just check for Arrow.jl which is required
        # and use memory-mapped files as a shared memory simulation
        isdefined(Main, :Arrow)
        return true
    catch
        return false
    end
end

# Global variable to track additional shared memory features
# Renamed to avoid duplicate declaration
global _ADDITIONAL_SHARED_MEMORY_FEATURES = check_shared_memory_available()

"""
    setup_shared_memory(cl::CrossLinkManager, data::Union{DataFrame, Arrow.Table}, identifier::String)

Set up shared memory for data and return info about the shared memory region.
"""
function setup_shared_memory(cl::CrossLinkManager, data, identifier::String)
    # Check if shared memory is available
    if !_SHARED_MEMORY_AVAILABLE
        cl.debug && @info "Shared memory not available"
        return nothing
    end
    
    try
        # Different methods to set up shared memory
        if _SHARED_MEMORY_MODULE == :CppWrapper && cl.cpp_available && cl.cpp_instance !== nothing
            try
                # Convert to Arrow Table if needed
                arrow_table = data isa Arrow.Table ? data : Arrow.Table(data)
                
                # Use C++ implementation
                shared_key = CrossLinkCppWrapper.setup_shared_memory(cl.cpp_instance, arrow_table, identifier)
                
                cl.debug && @info "Set up shared memory using C++ wrapper with key: $(shared_key)"
                return Dict("shared_memory_key" => shared_key, "module" => "cpp")
            catch e
                cl.debug && @warn "Failed to set up shared memory via C++: $e"
                # Continue to other methods instead of failing
            end
        elseif _SHARED_MEMORY_MODULE == :SharedArrays
            # Not implemented
            cl.debug && @info "SharedArrays shared memory setup not implemented"
        elseif _SHARED_MEMORY_MODULE == :Mmap
            # Not implemented
            cl.debug && @info "Mmap shared memory setup not implemented"
        end
        
        # If we got here, no method succeeded
        return nothing
    catch e
        cl.debug && @warn "Error setting up shared memory: $e"
        return nothing
    end
end

"""
    get_from_shared_memory(cl::CrossLinkManager, shared_memory_key::Union{String, Missing, Nothing})

Get an Arrow table from shared memory. If shared memory fails or key is missing, 
returns nothing so the caller can fall back to other methods.
"""
function get_from_shared_memory(cl::CrossLinkManager, shared_memory_key::Union{String, Missing, Nothing})
    # Handle missing or nothing case
    if shared_memory_key === missing || shared_memory_key === nothing
        cl.debug && @info "Shared memory key is missing or nothing, will fall back to standard queries"
        return nothing
    end
    
    # Check if shared memory is available
    if !_SHARED_MEMORY_AVAILABLE
        cl.debug && @info "Shared memory not available, will fall back to standard queries"
        return nothing
    end
    
    try
        # This is just a placeholder - actual shared memory implementation
        # would depend on OS and available libraries
        if _SHARED_MEMORY_MODULE == :SharedArrays
            # Use SharedArrays.jl approach
            cl.debug && @info "Using SharedArrays for shared memory access"
            # Not implemented
        elseif _SHARED_MEMORY_MODULE == :Mmap
            # Use memory mapping approach
            cl.debug && @info "Using Mmap for shared memory access"
            # Not implemented
        elseif _SHARED_MEMORY_MODULE == :CppWrapper
            # Use C++ wrapper
            cl.debug && @info "Using C++ wrapper for shared memory access"
            if cl.cpp_available && cl.cpp_instance !== nothing
                try
                    # Get the shared memory via C++ implementation
                    result = CrossLinkCppWrapper.get_shared_memory(cl.cpp_instance, shared_memory_key)
                    if result !== nothing
                        return result
                    else
                        cl.debug && @info "Shared memory key not found: $shared_memory_key, falling back to standard queries"
                    end
                catch e
                    cl.debug && @warn "Failed to get shared memory via C++: $e, falling back to standard queries"
                end
            else
                cl.debug && @info "C++ not available for shared memory access, falling back to standard queries"
            end
        end
        
        # If we got here, no method succeeded
        cl.debug && @info "No shared memory method succeeded for key: $shared_memory_key, falling back to standard queries"
        return nothing
    catch e
        cl.debug && @warn "Error accessing shared memory: $e, falling back to standard queries"
        return nothing
    end
end

"""
    fallback_to_duckdb(cl::CrossLinkManager, identifier::String)

Fallback to using DuckDB when shared memory access fails.
"""
function fallback_to_duckdb(cl::CrossLinkManager, identifier::String)
    cl.debug && @info "Falling back to DuckDB for identifier: $identifier"
    
    try
        # Connect to DuckDB
        con = DuckDB.connect(cl.db_path)
        
        # Query the data
        result = DBInterface.execute(con, "SELECT * FROM $identifier")
        
        # Convert to Arrow table
        return Arrow.Table(result)
    catch e
        cl.debug && @warn "Error in DuckDB fallback: $e"
        return nothing
    end
end

"""
    cleanup_all_shared_memory(db_path::String="crosslink.duckdb", force::Bool=false)

Clean up all shared memory files in the crosslink_shared directory.

This function can be called explicitly to clean up shared memory resources from previous runs
that might not have been properly deleted.

# Arguments
- `db_path`: Path to the DuckDB database file to determine the shared memory directory
- `force`: Force deletion even if files appear to be in use

# Returns
- Number of files removed
"""
function cleanup_all_shared_memory(db_path::String="crosslink.duckdb", force::Bool=false)
    try
        # Get the shared directory
        shared_dir = joinpath(dirname(abspath(db_path)), "crosslink_shared")
        
        # Check if directory exists
        if !isdir(shared_dir)
            return 0
        end
        
        # Count of removed files
        removed_count = 0
        
        # Remove all .shared files
        for filename in readdir(shared_dir)
            if endswith(filename, ".shared")
                file_path = joinpath(shared_dir, filename)
                try
                    # Check if safe to delete if force=false
                    if force || is_file_safe_to_delete(file_path)
                        rm(file_path)
                        removed_count += 1
                    end
                catch e
                    @warn "Failed to remove shared memory file $file_path: $e"
                end
            end
        end
        
        # Try to remove the directory if it's empty
        try
            if isempty(readdir(shared_dir))
                rm(shared_dir)
            end
        catch
            # Ignore errors when removing directory
        end
        
        return removed_count
    catch e
        @warn "Error cleaning up shared memory files: $e"
        return 0
    end
end

"""
    cleanup_shared_memory(cl::CrossLinkManager, shared_memory_key::String)

Clean up a shared memory region.
"""
function cleanup_shared_memory(cl::CrossLinkManager, shared_memory_key::Union{String, Missing, Nothing})
    # Handle missing or nothing case
    if shared_memory_key === missing || shared_memory_key === nothing
        return
    end
    
    # Check if shared memory is available
    if !_SHARED_MEMORY_AVAILABLE
        return
    end
    
    try
        # Different methods to clean up shared memory
        if _SHARED_MEMORY_MODULE == :CppWrapper && cl.cpp_available && cl.cpp_instance !== nothing
            try
                # Use C++ implementation
                CrossLinkCppWrapper.cleanup_shared_memory(cl.cpp_instance, shared_memory_key)
                cl.debug && @info "Cleaned up shared memory using C++ wrapper: $(shared_memory_key)"
            catch e
                cl.debug && @warn "Failed to clean up shared memory via C++: $e"
            end
        elseif _SHARED_MEMORY_MODULE == :SharedArrays
            # Not implemented
        elseif _SHARED_MEMORY_MODULE == :Mmap
            # Not implemented
        end
    catch e
        cl.debug && @warn "Error cleaning up shared memory: $e"
    end
end

"""
    safe_shared_memory_get(cl::CrossLinkManager, identifier::String)

Safely get data from shared memory with proper fallback to DuckDB.
"""
function safe_shared_memory_get(cl::CrossLinkManager, shared_memory_key::Union{String, Missing, Nothing}, identifier::String)
    # Try shared memory first
    result = get_from_shared_memory(cl, shared_memory_key)
    
    # If shared memory fails, fall back to DuckDB
    if result === nothing
        cl.debug && @info "Shared memory access failed for $identifier, falling back to DuckDB"
        result = fallback_to_duckdb(cl, identifier)
    end
    
    return result
end 